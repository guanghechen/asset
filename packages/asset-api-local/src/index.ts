export * from './AssetResolverApi'
export * from './AssetService'
export * from './types'
export * from './util/asset'
export * from './util/location'
