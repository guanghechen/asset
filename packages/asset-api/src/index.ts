export * from './AssetService'
export * from './AssetTask'
export * from './AssetTaskApi'
export * from './AssetDataCooker'
export * from './AssetTaskScheduler'
