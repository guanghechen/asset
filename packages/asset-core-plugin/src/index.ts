export * from './AssetParser'
export * from './types/plugin/plugin'
export * from './types/plugin/polish'
export * from './types/plugin/parse'
export * from './types/parser'
export * from './types/resolver'
export * from './util/misc'
