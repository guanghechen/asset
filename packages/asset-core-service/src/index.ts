export * from './AssetResolver'
export * from './types/misc'
export * from './util/asset'
export * from './util/hash'
export * from './util/TaskPipeline'
