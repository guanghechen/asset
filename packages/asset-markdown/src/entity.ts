import type { AssetDataItem } from '@guanghechen/site-api'

export const MarkdownAssetType = 'markdown'
// eslint-disable-next-line @typescript-eslint/no-redeclare
export type MarkdownAssetType = typeof MarkdownAssetType

/**
 * Markdown data
 */
export interface MarkdownAssetDataItem
  extends AssetDataItem<MarkdownAssetType> {}
