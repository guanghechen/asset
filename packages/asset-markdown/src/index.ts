export * from './entity'
export * from './processor'
