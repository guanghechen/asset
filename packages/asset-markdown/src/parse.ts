import gfm from 'remark-gfm'
import math from 'remark-math'
import markdown from 'remark-parse'
import unified from 'unified'
import removePosition from 'unist-util-remove-position'
import { AssetMarkdownEntityContent } from './entity'


export type MarkdownParser<T extends unknown = unknown> =
  (content: string) => AssetMarkdownEntityContent<T>


/**
 * Parse markdown to ast
 *
 * @param rawContent
 */
const processor = unified().use(markdown).use(gfm).use(math)
export const parseMarkdownToAst: MarkdownParser<any> = (rawContent) => {
  // resolve content
  const content: any = processor.runSync(processor.parse(rawContent))
  const summary: any = {
    type: 'root',
    children: content.children.slice(0, 3),
  }

  const result = {
    content: removePosition(content, false),
    summary: removePosition(summary, false),
  }
  return result
}


/**
 * Preserve the raw content
 *
 * @param rawContent
 */
export const preserveMarkdown: MarkdownParser<string> = (rawContent) => {
  return {
    content: rawContent,
    summary: '',
  }
}
