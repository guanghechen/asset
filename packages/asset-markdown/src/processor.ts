import dayjs from 'dayjs'
import yaml from 'js-yaml'
import invariant from 'tiny-invariant'
import {
  AssetProcessor,
  CategoryDataItem,
  ImmutableCategoryDataManager,
  ImmutableTagDataManager,
  RawCategoryDataItem,
  RawTagDataItem,
  RoughAssetDataItem,
  TagDataItem,
} from '@guanghechen/site-api'
import { AssetMarkdownEntity, AssetMarkdownType } from './entity'


/**
 * Props for building AssetMarkdownProcessor
 */
export interface AssetMarkdownProcessorProps {
  /**
   * Encoding of markdown files
   */
  encoding: BufferEncoding
  /**
   * Whether the meta data is optional
   */
  isMetaOptional?: boolean
  /**
   * Custom function to determine whether an asset file processable
   */
  processable?: AssetProcessor['processable']
}


/**
 * Processor for handle markdown asset
 */
export class AssetMarkdownProcessor implements AssetProcessor<AssetMarkdownEntity> {
  protected readonly encoding: BufferEncoding
  protected readonly isMetaOptional: boolean

  public constructor(props: AssetMarkdownProcessorProps) {
    const { encoding, isMetaOptional = true, processable } = props
    this.encoding = encoding
    this.isMetaOptional = isMetaOptional

    if (processable != null) {
      this.processable = processable.bind(this)
    }
  }

  /**
   * @override
   */
  public processable(filepath: string): boolean {
    return /\.md$/.test(filepath)
  }

  /**
   * @override
   */
  public process(
    filepath: string,
    _rawContent: Buffer,
    roughAsset: RoughAssetDataItem,
    tagDataManager: ImmutableTagDataManager,
    categoryDataManager: ImmutableCategoryDataManager,
  ): [AssetMarkdownEntity, TagDataItem[], CategoryDataItem[][]] {
    const rawContent: string = _rawContent.toString(this.encoding)
    let match: string[] | null = /^\s*[-]{3,}\n\s*([\s\S]*?)[-]{3,}\n/.exec(rawContent)

    if (this.isMetaOptional) {
      if (match == null) match = ['', '']
    } else {
      invariant(match != null, `No meta data found in ${ filepath }`)
    }

    const meta: Record<string, any> = yaml.safeLoad(match[1]) || {} as any
    const uuid: string = meta.uuid || roughAsset.uuid
    const title: string = meta.title || roughAsset.title
    const createAt = meta.createAt != null
      ? dayjs(meta.createAt).toDate().toISOString()
      : roughAsset.createAt
    const updateAt = meta.updateAt != null
      ? dayjs(meta.updateAt).toDate().toISOString()
      : roughAsset.updateAt

    // resolve tags
    const rawTags = (meta.tags || []) as RawTagDataItem[]
    const tags: TagDataItem[] = rawTags.map(rawTag => (
      tagDataManager.normalize(rawTag)
    ))

    // resolve categories
    const rawCategories = (meta.categories || []) as RawCategoryDataItem[][]
    const categories: CategoryDataItem[][] = rawCategories.map(categoryPath => (
      categoryPath.map(c => categoryDataManager.normalize(c))
    ))

    // resolve content
    const content: string = rawContent.slice(match[0].length).trim()

    const entity: AssetMarkdownEntity = {
      uuid,
      type: AssetMarkdownType,
      fingerprint: roughAsset.fingerprint,
      location: roughAsset.location,
      lastModifiedTime: roughAsset.lastModifiedTime,
      createAt,
      updateAt,
      title,
      tags: tags.map(tag => tag.uuid),
      categories: categories.map(cp => cp.map(c => c.uuid)),
      content,
      summary: '',
    }

    return [entity, tags, categories]
  }
}
