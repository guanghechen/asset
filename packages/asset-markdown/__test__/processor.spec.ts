import fs from 'fs-extra'
import path from 'path'
import {
  AssetDataManager,
  CategoryDataManager,
  RoughAssetDataItem,
  TagDataManager,
  sha1,
} from '@guanghechen/site-api'
import { AssetMarkdownProcessor } from '../src'

const caseRootDir = path.resolve(__dirname, 'cases')


describe('base', function () {
  test('processable', function () {
    const processor = new AssetMarkdownProcessor({ encoding: 'utf-8' })
    expect(processor.processable('/a/a.md')).toBeTruthy()
    expect(processor.processable('.../a/a.md')).toBeTruthy()
    expect(processor.processable('a.txt.md')).toBeTruthy()
    expect(processor.processable('a.md.txt')).toBeFalsy()

    const customProcessor = new AssetMarkdownProcessor({
      encoding: 'utf-8',
      processable: (filepath) => /\.(xmd|txt)$/.test(filepath),
    })
    expect(customProcessor.processable('/a/a.md')).toBeFalsy()
    expect(customProcessor.processable('.../a/a.md')).toBeFalsy()
    expect(customProcessor.processable('a.txt.md')).toBeFalsy()
    expect(customProcessor.processable('a.md.txt')).toBeTruthy()
    expect(customProcessor.processable('a.xmd')).toBeTruthy()
  })

  test('process', function () {
    function process(filepath: string) {
      const absoluteFilepath = path.resolve(caseRootDir, filepath)
      const rawContent = fs.readFileSync(absoluteFilepath)

      const createDate: Date = new Date('2020-01-01 00:00:00')
      const updateDate: Date = new Date('2020-01-02 12:34:56')
      const lastModifiedDate: Date = new Date('2020-01-02 12:34:56')

      const roughAsset: RoughAssetDataItem = {
        uuid: sha1(filepath),
        location: filepath,
        extname: path.extname(filepath),
        fingerprint: sha1(filepath),
        lastModifiedTime: lastModifiedDate.getMilliseconds(),
        createAt: createDate.toISOString(),
        updateAt: updateDate.toISOString(),
        title: filepath,
        tags: [],
        categories: [],
      }

      const tagDataManager = new TagDataManager('./tag.json')
      const categoryDataManager = new CategoryDataManager('./category.data.json')
      const assetDataManager = new AssetDataManager(caseRootDir, './asset.data.json')

      const processor = new AssetMarkdownProcessor({
        encoding: 'utf-8',
        isMetaOptional: true,
      })
      const result = processor.process(
        filepath,
        Buffer.from(rawContent),
        roughAsset,
        tagDataManager,
        categoryDataManager,
        assetDataManager,
      )
      return result.next()
    }

    expect(process('a.md')).toMatchSnapshot('a')
    expect(process('b.md')).toMatchSnapshot('b')
  })
})
