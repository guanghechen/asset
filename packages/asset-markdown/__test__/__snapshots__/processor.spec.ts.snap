// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base process: a 1`] = `
Array [
  Object {
    "categories": Array [],
    "content": "$E = m \\\\\\\\cdot c^2$",
    "createAt": "2019-12-31T16:00:00.000Z",
    "fingerprint": "5d48a79a3e66a0b67d03c582501493d256004f1f",
    "lastModifiedTime": 0,
    "location": "a.md",
    "tags": Array [],
    "title": "a.md",
    "type": "markdown",
    "updateAt": "2020-01-02T04:34:56.000Z",
    "uuid": "5d48a79a3e66a0b67d03c582501493d256004f1f",
  },
  Array [],
  Array [],
]
`;

exports[`base process: b 1`] = `
Array [
  Object {
    "categories": Array [
      Array [
        "a",
        "b",
      ],
      Array [
        "c",
        "d",
      ],
      Array [
        "e",
      ],
    ],
    "content": "$E = m \\\\\\\\cdot c^2$",
    "createAt": "2000-01-01T01:02:03.000Z",
    "fingerprint": "bb801cb4e571ec0bc7a6ad6d316dacf27c7b107a",
    "lastModifiedTime": 0,
    "location": "b.md",
    "tags": Array [
      "math",
      "algorithm",
    ],
    "title": "b.md",
    "type": "markdown",
    "updateAt": "1982-12-22T04:07:02.000Z",
    "uuid": "waw",
  },
  Array [
    Object {
      "assets": Array [],
      "title": "math",
      "uuid": "math",
    },
    Object {
      "assets": Array [],
      "title": "algorithm",
      "uuid": "algorithm",
    },
  ],
  Array [
    Array [
      Object {
        "assets": Array [],
        "children": Array [],
        "parents": Array [],
        "title": "A",
        "uuid": "a",
      },
      Object {
        "assets": Array [],
        "children": Array [],
        "parents": Array [],
        "title": "B",
        "uuid": "b",
      },
    ],
    Array [
      Object {
        "assets": Array [],
        "children": Array [],
        "parents": Array [],
        "title": "C",
        "uuid": "c",
      },
      Object {
        "assets": Array [],
        "children": Array [],
        "parents": Array [],
        "title": "D",
        "uuid": "d",
      },
    ],
    Array [
      Object {
        "assets": Array [],
        "children": Array [],
        "parents": Array [],
        "title": "E",
        "uuid": "e",
      },
    ],
  ],
]
`;
